Csfun graphs (#665)
Ada-Developers-Academy/textbook-curriculum: 04-cs-fundamentals/internship/graphs.md

### List of Edges

The most basic way to store a graph is simply to maintain a list of edges.  For example for the graph with numbers representing the nodes.

![Example graph for an edge list](images/graph-edge-list.png)

We could represent this graph with the following:

```ruby
sample_graph = [
  [1, 2], [1, 3], [3, 4]
]
```

This representation has a couple of advantages it's simple and uses a minim amount of space.  For a graph G(N, E) where N is the set of nodes and E the set of edges, this representation has a space complexity of O(E).

It does have a major downside, to find out if any particular pair of nodes are connected, you must traverse the entire list.  This means that the time complexity to find out of any two nodes are connected is O(E) as well.  We can do better.  Instead we will look at an adjacency matrix.

### Adjacentcy Matrix

Lets look at the same graph again.  Instead of maintaining a list of edges we could instead create a matrix where each row-column intersection indicates a potential edge between two nodes.

> ![Example graph for an edge list](images/graph-edge-list.png)

Reviewer: broken image?

Below is an example matrix:

> ![Example Adjacency matrix](images/graph-adjacency-matrix.png)

Reviewer: broken image? I got a 404

    @value = val
    @left = nil
    @right = nil
   end
end
```

In our methods we traversed the structure adding and removing elements by writing either iterative or recursive methods which jumped from one node to the next following the links between them.  We could write Node class for our graph algorithms like the example below.  In this example we have a potential `GraphNode`.  

```ruby
class GraphNode
  attr_reader :key, :value
  attr_accessor :edges

  def initialize(key, val, edges = [])
    @key = key
    @value = val
    @edges = edges
  end

  def add_connection(other_node)
    edges.push(other_node)
  end
end
```

While this can work it leaves us with a problem.  Unlike a tree or a linked list, there is no starting node for a graph.  So instead we would need another way to store our graph, which provides a method to start with any node.

### List of Edges

The most basic way to store a graph is simply to maintain a list of edges.  For example for the graph with numbers representing the nodes.

![Example graph for an edge list](images/graph-edge-list.png)

We could represent this graph with the following:

```ruby
sample_graph = [
  [1, 2], [1, 3], [3, 4]
]
```

> This representation has a couple of advantages it's simple and uses a minim amount of space.  For a graph G(N, E) where N is the set of nodes and E the set of edges, this representation has a space complexity of O(E).

Reviewer 2: ```suggestion
This representation has a couple of advantages it's simple and uses a minimum amount of space.  For a graph _G(N, E)_ where _N_ is the set of nodes and _E_ the set of edges, this representation has a space complexity of _O(E)_.
```

> It does have a major downside, to find out if any particular pair of nodes are connected, you must traverse the entire list.  This means that the time complexity to find out of any two nodes are connected is O(E) as well.  We can do better.  Instead we will look at an adjacency matrix.

Reviewer 2: ```suggestion
It does have a major downside, to find out if any particular pair of nodes are connected, you must traverse the entire list.  This means that the time complexity to find out of any two nodes are connected is _O(E)_ as well.  We can do better.  Instead we will look at an adjacency matrix.
```

### Adjacentcy Matrix

Lets look at the same graph again.  Instead of maintaining a list of edges we could instead create a matrix where each row-column intersection indicates a potential edge between two nodes.

![Example graph for an edge list](images/graph-edge-list.png)

Below is an example matrix:

![Example Adjacency matrix](images/graph-adjacency-matrix.png)

> In the adjacency matrix above if two nodes matched by the row-column intersection share an edge, that entry in the matri is true.  If they do not share an edge that entry is false.  Notice a couple of things:

Reviewer 2: ```suggestion
In the adjacency matrix above if two nodes matched by the row-column intersection share an edge, that entry in the matrix is true.  If they do not share an edge that entry is false.  Notice a couple of things:
```

- Unless a node has an edge with itself the diagonal from [0][0] to [n-1][n-1] will be false
- If the graph is undirected the entries across the central diagonal will be mirror images
  - if (1, 2) is true, then (2, 1) must also be true
- We would need some way to convert a node in the graph to a number

> For a graph G(N, E), where N is the number of nodes and E the number of edges, this solution provides:

Reviewer 2: ```suggestion
For a graph _G(N, E)_, where _N_ is the number of nodes and _E_ the number of edges, this solution provides:
```

- **O(1)** lookup time complexity to find out if any two nodes share an edge
- **O(N<sup>2</sup>)** space complexity to store this representation
- **O(1)** time complexity for getting a list of all the neighbors of a node `n` because you need to read the entire row or column.

This method provides a very fast lookup time and not terrible space complexity if the graph is a _dense graph_, in other words if there are a large number of edges compared to nodes.

> In the example above we used a matrix of booleans where true indicates an existing edge and false indicates no edge exists connecting the two nodes.  You will often see this represented as a matrix of ones and zeroes where zero indicates no edge exists and 1 indicates it does.  This has the advantage of only requiring one bit per matrix cell.  There are [some strategies](https://github.com/javolio/Bitwise-Adjacency-Matrix) to use this fact to compress the space requirements.  

Reviewer 2: ```suggestion
In the example above we used a matrix of booleans where true indicates an existing edge and false indicates no edge exists connecting the two nodes.  You will often see this represented as a matrix of ones and zeroes where zero indicates no edge exists and 1 indicates it does.  This has the advantage of only requiring one bit per matrix cell.  There are [some strategies](https://github.com/javolio/Bitwise-Adjacency-Matrix) to use this fact to compress the space requirements (though we still can't do better than _O(N<sup>2</sup>)_ bits).
```

### Adjacency List

An Adjacency List provides a bit of a middle ground between a list of edges and an adjacency matrix.  In an adjacency list, you have an array with each index representing a node and the content an unordered list of neighboring nodes.  The list can be an array or a linked list.

![Adjacency List Example](images/graph-adjacency-list.png)

This adjacency list gives us:

- **O(d)** lookup time to find out if any two nodes share an edge, where `d` is the _degree_ of a node meaning the number of edges the node has
  - The maximum value of d would be the case where a node is connected to every other node (assuming no duplicate edges) which is V - 1
  - This can be a problem in dense graphs
- O(2E) = **O(E)** space complexity to store this representation because each element of the array stores a list of the edges a node has, twice in the case of an undirected graph.

> - **O(1)** time complexity to find all the neighbors of a particular node because it is an O(1) operation to lookup a node by it's index and the list at that location is a list of all the node's neighbors.

Reviewer 2: ```suggestion
- **O(1)** time complexity to find list of all of the neighbors of a particular node (because it is an _O(1)_ operation to lookup a node by it's index and the list at that location is a list of all the node's neighbors).
```

**Pseudocode**

1. Start with a graph adjacency list and a specific node called `start_node`
1. Create a list of nodes called `visited` and mark each element `false`
1. Create an empty queue `q`
1. Add `start_node` to `q`
1. While `q` is not empty
    1. Remove an element from `q` and store it in `current`
    1. Mark `current` as `true` in the list `visited`
    1. Loop through each of `current`'s neighbors
        1. If the neighbor is not `true` in `visited` add it to `q`

You can see breadth-first-search animated on [HackerEarth](https://www.hackerearth.com/practice/algorithms/graphs/breadth-first-search/visualize/)

Breadth first search is a solution in a variety of problems including:

- Finding the shortest path in a graph/maze
- Solving puzzle games like a [Rubik's Cube](https://www.quora.com/How-can-solving-a-Rubiks-Cube-be-framed-as-a-graph-problem)
- Checking to see if a graph is connected

**Questions**

<details>
  <summary>What is the time complexity of Breadth-First-Search</summary>

>   Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is O(N + E) where `N` is the number of nodes in the graph and `E` is the number of edges since each node and each edge will be explored.

Reviewer 2: ```suggestion
  Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is _O(N + E)_ where _N_ is the number of nodes in the graph and _E_ is the number of edges since each node and each edge will be explored.
```

</details>

<details>
  <summary>What is the space complexity of Breadth-First-Search</summary>

>   In the worst-case you will need to add each node to the Queue, so the space complexity is O(N) where `N` is the number of nodes in the graph.

Reviewer 2: ```suggestion
  In the worst-case you will need to add each node to the Queue, so the space complexity is _O(N)_ where _N_ is the number of nodes in the graph.
```

</details>

### Depth-First-Search

Where breadth-first-search spreads out from a starting node in order of distance from the starting node, depth first search follows each path as far as possible before backing up and following the next closest path.

**Pseudocode**

1. Starting with a graph's adjacency list and a starting node `s`
1. Create a Stack called `stack`
1. Create a list of nodes called `visited` and mark each element `false`
1. Push `s` into `stack`
1. mark `s` as `true` in `visited`
1. while `stack` is not empty
    1. Remove an element from `stack` and store it in `current_node`
    1. Loop through all the neighbors of `current_node`
        1. If they are not marked as `true` in `visited`
            1. Push the neighbor into `stack`
            1. Mark the neighbor as `true` in `visited`

**Questions**

<details>
  <summary>What is the time complexity of Depth-First-Search</summary>

>   Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is O(N + E) where `N` is the number of nodes in the graph and `E` is the number of edges since each node and each edge will be explored.  Note, this is the same as breadth-first-search.

Reviewer 2: ```suggestion
  Since you will visit each node once, and loop through each of the edges in each node the Big-O of this algorithm is _O(N + E)_ where _N_ is the number of nodes in the graph and _E_ is the number of edges since each node and each edge will be explored.  Note that this is the same as for breadth-first-search.
```

</details>

<details>
  <summary>What is the space complexity of Depth-First-Search</summary>

>   In the worst-case you will need to add each node to the Stack, so the space complexity is O(N) where `N` is the number of nodes in the graph.

Reviewer 2: ```suggestion
  In the worst-case you will need to add each node to the Stack, so the space complexity is _O(N)_ where _N_ is the number of nodes in the graph.  Note that this is the same as for breadth-first-search.
```

        nearest_node  = node_index
        shortest_distance = shortest_distances[node_index]
      end
    end  

    # Mark the picked vertex as visited
    added[nearest_node] = true
    # Update dist value of the
    # adjacent nodes of the picked node.
    (0...num_nodes).each do |node_index|
      edge_distance = adjacency_matrix[nearest_node][node_index]

      if (edge_distance > 0 && 
          ((shortest_distance + edge_distance) <  
              shortest_distances[node_index]))  
        parents[node_index] = nearest_node
        shortest_distances[node_index] = shortest_distance +  
                                        edge_distance
      end
    end
  end
  return {  
    start_node: start_node,
    parent_list: parents,
    shortest_distances: shortest_distances
  }
end
```

**Questions**

<details>
  <summary>Is Dijkstra's Algorithm a greedy algorithm?</summary>
  <strong>YES!</strong>  With each iteration of the main loop Dijkstra's algorithm picks the next node with the shortest path.  It makes each choice by picking the next shortest path available.  
</details>

<details>
  <summary>What is the time complexity of Dijkstra's Algorithm?</summary>

>   <strong>O(N<sup>2</sup>)</strong>  Where N is the number of nodes.  The algorithm iterates until each node has been visited, N times, and with each iteration you have to find the node with the smallest current distance, so with each iteration you have to look roughly at most N other nodes.

Reviewer 2: ```suggestion
  _O(N<sup>2</sup>)_  Where _N_ is the number of nodes.  The algorithm iterates until each node has been visited, N times, and with each iteration you have to find the node with the smallest current distance, so with each iteration you have to inspect at most _N_ other nodes.
```

</details>

**Exercise** 

Take a look at the solution above.  Can you convert it to a solution involving an Adjacency list?  Each entry in the adjacency list will need to store a weight as well as the destination node.

#### Using a Heap

> Because finding an unvisited node with a minimum distance takes O(N) time, Dijkstra's algorithm can be made more efficient by using a Min-Heap where nodes are stored by order of their distance.  

Reviewer 2: ```suggestion
Because finding an unvisited node with a minimum distance takes _O(N)_ time, Dijkstra's algorithm can be made more efficient by using a Min-Heap where nodes are stored by order of their distance.  
```

By using an Adjacency list instead of an Adjacency matrix and using a heap to store the distances to each node you can get:

> - Finding the next node to visit using the heap is O(log N)

Reviewer 2: ```suggestion
- Finding the next node to visit using the heap is _O(log N)_
```

> - With an adjacency list, updating the distances to each node takes O(log N) each with a total of E edges.

Reviewer 2: ```suggestion
- With an adjacency list, updating the distances to each node takes _O(log N)_ each with a total of _E_ edges.
```

> - Overall O(N log N + E log N) = O( (N + E) log N)

Reviewer 2: ```suggestion
- Overall _O(N log N + E log N)_ = _O((N + E) log N)_
```

#### Limitation of Dijkstra's Algorithm

Can you see a problem with Dijkstra's algorithm?  What if an edge has a _negative_ weight?  Then you might find a shorter path **after** already visiting a node.  This is often not a problem, but it is a limitation of the algorithm.

For example you could build a graph of chemical reactions between compounds.  The edge weight could represent the energy released or absorbed by the reaction.

## Summary

In this lesson we have looked at the Graph data structure.  A graph is a collection of nodes and connections between nodes called edges.  Binary search trees and linked lists are both subsets of graphs.  A graph can have both unweighted and weighted edges.  In a weighted graph, each connection is assigned a cost or weight.  Further a graph can be directed or undirected.  In a directed graph an edge is not bidirectional.  

> There are two main ways to represent a graph.  The first is using an adjacency matrix which is a 2-dimensional array where each entry represents an edge between the nodes indicated by it's row and column indices.  An adjacency matrix is very quick to lookup a connection between two nodes.  On the other hand retrieving a list of all the edges a node has a time complexity of O(N) where `N` is the number of nodes.  It also require O(N<sup>2</sup>) space complexity.  The other method of representing a graph involves an adjacency list, which is an array of lists of neighbors.  This method has an O(1) time complexity to look up all the neighbors of a node, but O(d) to lookup a particular connection where `d` is the degree of the node, or the number of neighbors it has.  It also requires less space complexity.

Reviewer 2: ```suggestion
There are two main ways to represent a graph.  The first is using an adjacency matrix which is a 2-dimensional array where each entry represents an edge between the nodes indicated by it's row and column indices.  An adjacency matrix is very quick to lookup a connection between two nodes.  On the other hand retrieving a list of all the edges a node has a time complexity of _O(N)_ where _N_ is the number of nodes.  It also require _O(N<sup>2</sup>)_ space complexity.  The other method of representing a graph involves an adjacency list, which is an array of lists of neighbors.  This method has an _O(1)_ time complexity to look up all the neighbors of a node, but _O(d)_ to lookup a particular connection where _d_ is the degree of the node, or the number of neighbors it has.  It also requires less space complexity when the graph is sparse.
```

There are a number of algorithms involving graphs in computer science.  To traverse a graph you can either make a breadth-first-search or a depth-first-search.  In a breadth-first-search each node is investigated by it's proximity to the starting node.  In a depth-first-search a path is explored as far as it will go, and then the algorithm backs up until another subpath can be explored, like a mouse exploring a maze.

In weighted graphs, finding the shortest paths is a common problem.  One classic solution is Dijkstra's algorithm which is a good example of a greedy algorithm.

## Terms & Definitions

| Term | Definition |
|--- |--- |
| Graph | A data structure with two parts:  A limited set of verticies, called nodes, and a limited set of connections between nodes known as edges. |

> | Edge | A connection between two nodes in a graph |

Reviewer 2: Most of these have periods.

```suggestion
| Edge | A connection between two nodes in a graph. |
```

| Node | The fundamental unit of a graph in computer science.  Each element of a graph, connected by edges. |
| Sparse Graph | A graph with few edges when compared to nodes. |
| Dense Graph | A graph with many more edges when compared to nodes.  A graph is considered _dense_ when it has close to the maximum number of possible edges. |
| Directed Graph |  A graph where the edges connet in one direction.  For example an edge may conect nodes A -> B, but will not connect B -> A. |

> | Undirected Graph | A Graph in which each edge is bidirectional.  So an edge from A -> B also connects B -> A |

Reviewer 2: ```suggestion
| Undirected Graph | A Graph in which each edge is bidirectional.  So an edge from A → B also connects B → A. |
```

> | Adjacency Matrix | A 2-dimensional array listing the connections between nodes in a graph |

Reviewer 2: ```suggestion
| Adjacency Matrix | A 2-dimensional array listing the connections between nodes in a graph. |
```

| Adjacency List | A collection of unordered lists used to represent a graph.  Each list describes the neighors of a particular node.

> | Node Degree | The number of edges a particular node has |

Reviewer 2: ```suggestion
| Node Degree | The number of edges a particular node has. |
```

## Resources

- [BaseCS on Graphs](https://medium.com/basecs/from-theory-to-practice-representing-graphs-cfd782c5be38)

> - [Dijkstra's Algorithm]()

Reviewer 2: ```suggestion
- [Wikipedia: Dijkstra's Algorithm](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm)
```
